wai/src/AWS/Lambda/Wai/Encoding.hs

```diff
@@ -0,0 +1,208 @@
+gwReqToWai
+  :: HTTPRequest
+  -> IO W.Request
+gwReqToWai r@HTTPRequest{..} = do
+  remoteHost <- getRemoteHost
+  isFirstRef <- newIORef False
+
+  let  requestBody = do
+         isFirst <- atomicModifyIORef' isFirstRef (False,)
+         pure $ if isFirst then requestBodyData else B.empty
+
+       requestBodyLength = W.KnownLength (fromIntegral (B.length requestBodyData))
+
+  pure $ W.Request{..}
+
+  where
+    getRemoteHost = do
+      let srcIp = _requestContext ^. http . sourceIp . to T.unpack
+      addrs <- Sock.getAddrInfo Nothing (Just srcIp) Nothing
+      maybe (throwIO . RemoteHostException $ "Failed to get addrinfo for " <> srcIp)
+            (pure . Sock.addrAddress)
+            (listToMaybe addrs)
+
+    requestBodyData :: B.ByteString
+    requestBodyData = either (const B.empty) id $ requestBodyBytes r
+
+    requestMethod = TE.encodeUtf8 $ _requestContext ^. http . method
+    httpVersion = H.http10
+    rawPathInfo = TE.encodeUtf8 _rawPath
+    rawQueryString = TE.encodeUtf8 _rawQueryString
+    requestHeaders = ((CI.mk . TE.encodeUtf8) *** TE.encodeUtf8) <$> Map.toList _headers
+    isSecure = _requestContext ^. http . protocol . to (== "https")
+    pathInfo = H.decodePathSegments rawPathInfo
+    queryString = H.parseQuery rawQueryString
+    vault = mempty
+    requestHeaderHost = _requestContext ^. domainName . to (Just . TE.encodeUtf8)
+    requestHeaderRange = Nothing
+    requestHeaderReferer = List.lookup H.hReferer requestHeaders
+    requestHeaderUserAgent = _requestContext ^. http . userAgent . to (Just . TE.encodeUtf8)
+
+
+
+
+{- | Converts WAI HTTP response to API GW response
+     Note that API GW only truly supports text data,
+     i.e. except for response body (which can be base64 encoded)
+     all headers, etc must be valid UTF8.
+
+     Also note that text and JSON content types are assumed to be
+     valid UTF8-encoded binary strings.
+-}
+waiResToGw
+  :: W.Response
+  -> IO HTTPResponse
+waiResToGw waiRes = do
+  (resBody, resBase64) <- resBodyEncoding
+
+  pure $ HTTPResponse
+    { _statusCode = H.statusCode $ W.responseStatus waiRes
+    , _headers = resHeaders -- TODO: confirm that this works for multi-value headers and cookies
+    , _multiValueHeaders = Nothing -- TODO: confirm that 'headers' is enough
+    , _body = resBody
+    , _isBase64Encoded = resBase64
+    , _cookies = Nothing  -- TODO: confirm that 'headers' is enough
+    }
+
+  where
+    -- not using a map because here the comparison must be case-insensitive but
+    -- the list is usually tiny
+    resContentType :: Maybe B.ByteString
+    resContentType
+      = List.lookup H.hContentType
+      $ W.responseHeaders waiRes
+
+
+    resHeaders :: Maybe (Map Text Text)
+    resHeaders = case W.responseHeaders waiRes
+      of [] -> Nothing
+         hs -> Just . Map.fromList $
+               (    TE.decodeUtf8 . CI.original
+               *** TE.decodeUtf8
+               ) <$> hs
```

Comment:
If you join duplicates with a comma, you'll preserve semantics for everything except `Set-Cookie`.  But [this multiValueHeaders guide](https://aws.amazon.com/blogs/compute/support-for-multi-value-parameters-in-amazon-api-gateway/) makes me think that maybe you should be using that, and leave this and cookies empty.

Reply:
Yep, you are right. I was going to test that but re-reading it today, of course, since it's a map it'll lose dup fields. Will update to use that other field exclusively. 
The reason I went with this one first is because I didn't want to parse fields that are already multi-value when they come out of WAI ... but I'll test that instead. Maybe it just works.